core/src/main/java/org/apache/iceberg/TableMetadataBuilder.java

```diff
@@ -0,0 +1,252 @@
+    this.lastAssignedPartitionId = inputLastAssignedPartitionId;
+    return this;
+  }
+
+  public TableMetadataBuilder withDefaultSortOrderId(int inputDefaultSortOrderId) {
+    this.defaultSortOrderId = inputDefaultSortOrderId;
+    return this;
+  }
+
+  public TableMetadataBuilder withSortOrders(List<SortOrder> inputSortOrders) {
+    this.sortOrders = inputSortOrders;
+    return this;
+  }
+
+  public TableMetadataBuilder withProperties(Map<String, String> inputProperties) {
+    this.properties = inputProperties;
+    return this;
+  }
+
+  public TableMetadataBuilder withCurrentSnapshotId(long inputCurrentSnapshotId) {
+    this.currentSnapshotId = inputCurrentSnapshotId;
+    return this;
+  }
+
+  public TableMetadataBuilder withSnapshots(List<Snapshot> inputSnapshots) {
+    this.snapshots = inputSnapshots;
+    return this;
+  }
+
+  public TableMetadataBuilder withSnapshotLog(List<HistoryEntry> inputSnapshotLog) {
+    this.snapshotLog = inputSnapshotLog;
+    return this;
+  }
+
+  public TableMetadataBuilder withPreviousFiles(List<TableMetadata.MetadataLogEntry> inputPreviousFiles) {
+    this.previousFiles = inputPreviousFiles;
+    return this;
+  }
+
+  public TableMetadataBuilder from(TableMetadata base) {
+    this.formatVersion = base.formatVersion();
+    this.uuid = base.uuid();
+    this.location = base.location();
+    this.lastSequenceNumber = base.lastSequenceNumber();
+    this.lastUpdatedMillis = base.lastUpdatedMillis();
+    this.lastColumnId = base.lastColumnId();
+    this.currentSchemaId = base.currentSchemaId();
+    this.schemas = base.schemas();
+    this.specs = base.specs();
+    this.defaultSpecId = base.defaultSpecId();
+    this.lastAssignedPartitionId = base.lastAssignedPartitionId();
+    this.defaultSortOrderId = base.defaultSortOrderId();
+    this.sortOrders = base.sortOrders();
+    this.properties = base.properties();
+    this.currentSnapshotId = base.currentSnapshotId();
+    this.snapshots = base.snapshots();
+    this.snapshotLog = base.snapshotLog();
+
+    this.baseFormatVersion = base.formatVersion();
+    this.baseFile = base.file();
+    this.baseLastUpdateMillis = base.lastUpdatedMillis();
+    this.basePreviousFiles = base.previousFiles();
+    return this;
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  public TableMetadata build() {
+    if (formatVersion == null) {
+      formatVersion = TableMetadata.DEFAULT_TABLE_FORMAT_VERSION;
+    } else if (formatVersion != baseFormatVersion) {
+      Preconditions.checkArgument(formatVersion <= TableMetadata.SUPPORTED_TABLE_FORMAT_VERSION,
+          "Cannot upgrade table to unsupported format version: v%s (supported: v%s)",
+          formatVersion, TableMetadata.SUPPORTED_TABLE_FORMAT_VERSION);
+      Preconditions.checkArgument(formatVersion >= baseFormatVersion,
+          "Cannot downgrade v%s table to v%s", baseFormatVersion, formatVersion);
+    }
```

Comment:
I generally prefer to structure builders so that validation is done as soon as possible. That keeps `build` methods smaller. This could be moved into the `withFormatVersion` method.